README.md
<p align="center">
  <a href="https://wasiqb.github.io/coteafs/appium/">
    <img src="assets/coteafs-appium-logo.png" width=300 padding=10 />
  </a>
</p>

<h1 align="center">Wrapper Appium Framework which supports Automation of Mobile and Tablet apps for Android and iOS Real Devices / Emulators.</h1>

<div align="center">

> **Note:** Documentation site is under updation. Although some part of documentation is ready.
> Checkout this space to know when the documentation is ready.

[![Open Source Love](https://badges.frapsoft.com/os/v1/open-source.svg?v=103)][home]
[![CircleCI](https://circleci.com/gh/WasiqB/coteafs-appium.svg?style=svg)][circleci]
[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/WasiqB/Coteafs-Appium.svg)](http://isitmaintained.com/project/WasiqB/Coteafs-Appium "Average time to resolve an issue")
[![Bugs](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=bugs)](https://sonarcloud.io/project/issues?id=com.github.wasiqb.coteafs%3Aappium&resolved=false)
[![Test Coverage](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=coverage)](https://sonarcloud.io/component_measures?id=com.github.wasiqb.coteafs%3Aappium&metric=Coverage)
[![Quality Gate](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=alert_status)](https://sonarcloud.io/dashboard?id=com.github.wasiqb.coteafs%3Aappium)
[![Maintainability](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=sqale_rating)](https://sonarcloud.io/component_measures?id=com.github.wasiqb.coteafs%3Aappium&metric=Maintainability)
[![Reliability](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=reliability_rating)](https://sonarcloud.io/component_measures?id=com.github.wasiqb.coteafs%3Aappium&metric=Reliability)
[![Security](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=security_rating)](https://sonarcloud.io/component_measures?id=com.github.wasiqb.coteafs%3Aappium&metric=Security)
[![Vulnurability](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=vulnerabilities)](https://sonarcloud.io/component_measures?id=com.github.wasiqb.coteafs%3Aappium&metric=new_vulnerabilities)
[![Duplicate Code](https://sonarcloud.io/api/project_badges/measure?project=com.github.wasiqb.coteafs%3Aappium&metric=duplicated_lines_density)](https://sonarcloud.io/component_measures?id=com.github.wasiqb.coteafs%3Aappium&metric=Duplications)
[![Maven Central](https://img.shields.io/maven-central/v/com.github.wasiqb.coteafs/appium.svg)][maven]
[![Github Releases](https://img.shields.io/github/downloads/WasiqB/coteafs-appium/total.svg)](https://github.com/WasiqB/coteafs-appium/releases)

</div>

<p align="center">
  <img src="assets/android-test.gif" width=700 padding=10 style="border: 2px, border-color: black" />
</p>

## :rocket: [Quick Start][wiki]
* The [documentations][wiki] of coteafs-appium includes all the information you need to get started including setup, usage, advantages, sample test.
* To know what changes are Released, check out the [change log][] for complete list of changes.
* Want to know when our next feature or fix release is going to happen? Watch out our planned [milestones][].

## :factory: [How to configure the tests?][config-doc]

First step in writing tests using coteafs-appium framework is defining a Yaml config file in the `src/test/resources` folder.

**(_For more details, check the link above._)**

Sample file is shown below.

```yaml
servers:
  android:
    host: 127.0.0.1
    port: 4723
    external: true
    arguments:
      log_level: DEBUG
      log_time_stamp: true
      local_time_zone: true
      session_override: true
      android:
        suppress_adb_kill_server: true
devices:
  test:
    platform_type: ANDROID
    device_name: MI Redmi Note 4
    device_version: 7.0
    app_type: HYBRID
    device_type: REAL
    automation_name: APPIUM
    app_location: apps/android/VodQA.apk
    session_timeout: 120000
    playback:
      delay_before_swipe: 200
      delay_after_swipe: 100
      delay_before_tap: 0
      delay_after_tap: 0
```

## :dart: [How simple it is to write the tests?][test-doc]

By using Action classes for each Activity, the flow specific for that activity can be modularized and tests looks much clean and readable.
See the sample test below.

```java
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;

import com.github.wasiqb.coteafs.appium.android.vodqa.actions.LoginActivityAction;
import com.github.wasiqb.coteafs.appium.service.AppiumServer;

public class SampleTest {
  protected AndroidDevice androidDevice;
  private AppiumServer androidServer;

  @BeforeClass
  public void setupTestSuite () {
    this.androidServer = new AppiumServer ("android");  // Here the parameter refers to the key in server block in config file.
    this.androidServer.start ();

    this.androidDevice = new AndroidDevice (this.androidServer, "test");  // Here the param refers to the key in devices block in config file.
    this.androidDevice.start ();
  }

  @AfterClass (alwaysRun = true)
  public void tearDownTestSuite () {
    this.androidDevice.stop ();
    this.androidServer.stop ();
  }

  @Test
  public void login () {
    final LoginActivityAction login = new LoginActivityAction (this.androidDevice);
    login.addInputValue ("UserName", "admin")
      .addInputValue ("Password", "admin")
      .perform ();
  }
}
```

## :golf: [How to create Activity class?][activity-doc]

New class needs to be created for each Activity. There's an abstract activity class for each type of device.
Here `prepare` method needs to be implemented with all the elements available on that Activity.

See the below sample for an Android activity with comments for better understanding.

```java
import org.openqa.selenium.By;

import com.github.wasiqb.coteafs.appium.android.AndroidDevice;
import com.github.wasiqb.coteafs.appium.device.DeviceElement;

public class LoginActivity extends AndroidActivity {
  public LoginActivity (final AndroidDevice device) {
    super (device);
  }

  @Override
  protected DeviceElement prepare () {
    final DeviceElement main = DeviceElement.create ("Main")
      .forAndroid (By.id ("android:id/content"));
    DeviceElement.create ("Back")
      .parent (main)
      .forAndroid (By.xpath ("//android.widget.TextView[@text=\"Back\"]"))
      // We can set multiple locators for different Automation names.
      .forAndroid (AutomationType.UIAUTOMATOR2, MobileBy.AndroidUIAutomator ("new UiSelector ().text (\"Back\");"));
    DeviceElement.create ("UserName")
      .forAndroid (MobileBy.AccessibilityId ("username"))
      .parent (main);
    DeviceElement.create ("Password")
      .forAndroid (MobileBy.AccessibilityId ("password"))
      .parent (main);
    DeviceElement.create ("Login")
      .index (1)    // Index of element when multiple elements for same locator exists.
      .waitStrategy (WaitStrategy.VISIBLE)  // Wait strategy to be used while finding the element.
      .forAndroid (MobileBy.AccessibilityId ("login"))     // Locator used to find the element.
      .parent (main);    // Parent of current element.
    return main;
  }
}
```

## :soccer: [How to create your Activity action class?][action-doc]

There is abstract action class provided by framework where Activity specific flow is implemented in perform method.
See the sample Activity action class below.

```java
import com.github.wasiqb.coteafs.appium.android.AndroidActivityActions;
import com.github.wasiqb.coteafs.appium.android.AndroidDevice;
import com.github.wasiqb.coteafs.appium.android.vodqa.activities.LoginActivity;

public class LoginActivityAction extends AndroidActivityActions {
  public LoginActivityAction (final AndroidDevice device) {
    super (device);
  }

  @Override
  public void perform () {
    final LoginActivity login = new LoginActivity (getDevice ());
    login.onElement ("UserName")
      .enterText (value ("UserName"));
    login.onElement ("Password")
      .enterText (value ("Password"));
    login.onDevice ()
      .hideKeyboard ();
    login.onElement ("Login")
      .tap ();
  }
}
```

## :+1: Liked the framework until now?
### Add dependency to your projects POM.

```xml
<dependency>
  <groupId>com.github.wasiqb.coteafs</groupId>
  <artifactId>appium</artifactId>
  <version>3.0.0</version>
</dependency>
```

## :pushpin: [Want to know Key Features?][intro-doc]
In order to use a framework, it's important to know it's advantages. Let's see what are the key features of this framework:
* :iphone: Supports Android and iOS Real Devices and Emulators.
* :computer: Able to start and stop server on run-time and also can connect to already running server.
* :notebook: Enforces Page object model style of coding.
* :milky_way: Allows parallel and sequential execution of tests.
* :hammer: All capabilities, playback and delay settings are configurable through config file.
* :cloud: Supports execution of tests on any Cloud solution like BrowserStack, SauceLabs, TestingBot, etc.
* :video_camera: Supports video recording of tests on Android and iOS.
* :camera: Supports capturing screenshots for Android and iOS.
* :clipboard: Supports reading Clipboard from devices.
* :notebook_with_decorative_cover: Supports logging of all events occurred during test execution.
* :x: Provides pre-defined errors which wraps the Appium exceptions in a meaningful way.

CHANGELOG.md
## Version-3.0.0-beta1 - Coming soon
- **[BREAKING CHANGE]** [#38][38] : Added support to add multiple locators depending on Platform and Automation type.
  - Removed method `using` from `DeviceElement` and added `forAndroid` and forIos` methods to capture locators for Android and iOS platforms.
- **[ENHANCEMENT]** [#50][50] : Support for cloud based solution like BrowserStack, SauceLabs, etc.
- **[ENHANCEMENT]** [#54][54] : Updated README.
- **[ENHANCEMENT]** [#60][60] : Battery check is added. No action will be performed and execution will fail if Battery level is less than 20%.
- **[ENHANCEMENT]** [#56][56] :  Video record tests for Android and iOS.
- **[ENHANCEMENT]** [#63][63] : Common settings block for device and server section in config file.
- **[ENHANCEMENT]** [#75][75] : New missing actions added:
  - [x] Double tap
  - [x] N Tap
  - [x] Get Orientation
  - [x] Set Orientation
  - [x] Toggle Wifi (_Only Android_)
  - [x] Toggle Data (_Only Android_)
  - [x] Toggle Airplane (_Only Android_)
  - [x] Toggle Location service (_Only Android_)
  - [x] Get Clipboard (Text, Image, URL)
- **[FIX]** [#41][41] : Refactoring for latest Appium Java client.
- **[FIX]** [#42][42] : Build moved from Travis CI to Circle CI.
- **[FIX]** [#62][62] : Port handling for SauceLabs.
- **[FIX]** [#65][65] : NullPointerException when some keys missing from config.
- **[FIX]** [#71][71] : Error when only package name given instead of `app_location` config key.
- **[FIX]** [#78][78] : Action class not support for iOS, gestures handled with commands.
- **[UPGRADE]** [#51][51] : Upgraded dependencies:
  - **coteafs-parent:** 1.5.0 -> 2.1.0
  - **coteafs-config:** 1.5.0 -> 1.7.0
  - **coteafs-error:** 1.4.0 -> 1.6.0
  - **coteafs-logger:** 1.5.0 -> 1.7.0
  - **selenium-java:** 3.10.0 -> 3.141.5
  - **appium-java:** 5.0.4 -> 6.1.0

## Version-2.2.0 - 16/03/2018
### Missing Features:
* [#23][23]: UiAutomator is not supported in the framework.
* [#24][24]: Swipe on element does not work.
* [#37][37]: No availability to press enter key after entering search text in the search bar of the app.

### Fixes:
* [#26][26], [#28][28], [#29][29], [#30][30], [#31][31], [#32][32], [#33][33], [#34][34], [#35][35], [#36][36]: Code Climate issues.

### Dependencies Update:
* **coteafs-parent:** 1.4.0 -> 1.5.0
* **selenium-java:** 3.9.1 -> 3.10.0
* **coteafs-config:** 1.4.0 -> 1.5.0
* **coteafs-error:** 1.3.0 -> 1.4.0
* **coteafs-logger:** 1.4.0 -> 1.5.0

> Sample config file attached with release for reference.

## Version-2.1.0 - 08/12/2017
### Enhancements
* [#20][20] - Log file should be backedup and new logs should be created on every run.
* [#21][21] - Screenshot should be captured on UI action failure.

### Bug fixes
* [#19][19] - Error displayed in the console is not captured in log file.
* [#22][22] - Unable to run the tests using Chrome browser on Android mobile.

## Version-2.0.0 - 09/11/2017
* [#1][1] **- [NEW] -** Add unit tests for the Framework for Android.
* [#5][5] **- [NEW] -** Allow multiple inline verification on an element.
* [#9][9] **- [NEW] -** Required:New feature in the Framework - Dynamic Screenshot Naming Convention.
* [#11][11] **- [NEW] -** Incorporate latest Appium Java client v-5.0.0 for device level actions.
* [#12][12] **- [NEW] -** Provide support for Windows devices.
* [#13][13] **- [FIXED] -** Error "Missing artifact com.github.wasiqb.coteafs:configs:jar:${coteafs.config.version}"
* [#14][14] **-[NEW] -** Require configurable delay between user actions.
* [#15][15] **- [NEW] -** New Abstract class for Activity Actions needs to be added.
* [#16][16] **- [NEW] -** In config file, port should be made optional.
* [#17][17] **- [NEW] -** Need support for multiple Server Arguments.

## Version-1.0.0 - 27/08/2017
* [#2][2] **- [NEW] -** Provide support to Mobile Web App Automation for Android and iOS
* [#3][3] **- [NEW] -** Provide support for executing run on simulators / emulators for Android and iOS.
* [#7][7] **- [NEW] -** Incorporate coteafs-config project.
* [#8][8] **- [NEW] -** Feature to handle dynamic elements like lists and tables.
* [#10][10] **- [NEW] -** Allow configurable logging.
* [#4][4] **- [FIXED] -** No proper Exception when name is incorrect in method onElement.
* [#6][6] **- [FIXED] -** Solve issues in code identified by SonarQube.

## Initial Version 0.0.1
See README file for details.

[1]: https://github.com/WasiqB/coteafs-appium/issues/1
[2]: https://github.com/WasiqB/coteafs-appium/issues/2
[3]: https://github.com/WasiqB/coteafs-appium/issues/3
[4]: https://github.com/WasiqB/coteafs-appium/issues/4
[5]: https://github.com/WasiqB/coteafs-appium/issues/5
[6]: https://github.com/WasiqB/coteafs-appium/issues/6
[7]: https://github.com/WasiqB/coteafs-appium/issues/7
[8]: https://github.com/WasiqB/coteafs-appium/issues/8
[9]: https://github.com/WasiqB/coteafs-appium/issues/9

src/main/java/com/github/wasiqb/coteafs/appium/device/Device.java
/**
 * Copyright (c) 2017, Wasiq Bhamla.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.wasiqb.coteafs.appium.device;

import static com.github.wasiqb.coteafs.appium.constants.ConfigKeys.COTEAFS_CONFIG_DEFAULT_FILE;
import static com.github.wasiqb.coteafs.appium.constants.ConfigKeys.COTEAFS_CONFIG_KEY;
import static com.github.wasiqb.coteafs.appium.constants.ErrorMessage.SERVER_STOPPED;
import static com.github.wasiqb.coteafs.appium.utils.CapabilityUtils.setCapability;
import static com.github.wasiqb.coteafs.appium.utils.ErrorUtils.fail;
import static com.github.wasiqb.coteafs.appium.utils.ScreenRecorder.saveRecording;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.ADB_PORT;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.ANDROID_INSTALL_TIMEOUT;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.APP_ACTIVITY;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.APP_PACKAGE;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.APP_WAIT_ACTIVITY;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.APP_WAIT_DURATION;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.APP_WAIT_PACKAGE;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.AUTO_GRANT_PERMISSIONS;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.AVD;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.AVD_LAUNCH_TIMEOUT;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.AVD_READY_TIMEOUT;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.CHROMEDRIVER_EXECUTABLE;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.DEVICE_READY_TIMEOUT;
import static io.appium.java_client.remote.AndroidMobileCapabilityType.SYSTEM_PORT;
import static io.appium.java_client.remote.IOSMobileCapabilityType.APP_NAME;
import static io.appium.java_client.remote.IOSMobileCapabilityType.AUTO_ACCEPT_ALERTS;
import static io.appium.java_client.remote.IOSMobileCapabilityType.AUTO_DISMISS_ALERTS;
import static io.appium.java_client.remote.IOSMobileCapabilityType.BUNDLE_ID;
import static io.appium.java_client.remote.IOSMobileCapabilityType.LAUNCH_TIMEOUT;
import static io.appium.java_client.remote.IOSMobileCapabilityType.SHOW_XCODE_LOG;
import static io.appium.java_client.remote.IOSMobileCapabilityType.UPDATE_WDA_BUNDLEID;
import static io.appium.java_client.remote.IOSMobileCapabilityType.USE_NEW_WDA;
import static io.appium.java_client.remote.IOSMobileCapabilityType.USE_PREBUILT_WDA;
import static io.appium.java_client.remote.IOSMobileCapabilityType.WDA_CONNECTION_TIMEOUT;
import static io.appium.java_client.remote.IOSMobileCapabilityType.WDA_LOCAL_PORT;
import static io.appium.java_client.remote.IOSMobileCapabilityType.WDA_STARTUP_RETRIES;
import static io.appium.java_client.remote.IOSMobileCapabilityType.WDA_STARTUP_RETRY_INTERVAL;
import static io.appium.java_client.remote.IOSMobileCapabilityType.XCODE_ORG_ID;
import static io.appium.java_client.remote.IOSMobileCapabilityType.XCODE_SIGNING_ID;
import static io.appium.java_client.remote.MobileCapabilityType.APP;
import static io.appium.java_client.remote.MobileCapabilityType.AUTOMATION_NAME;
import static io.appium.java_client.remote.MobileCapabilityType.CLEAR_SYSTEM_FILES;
import static io.appium.java_client.remote.MobileCapabilityType.DEVICE_NAME;
import static io.appium.java_client.remote.MobileCapabilityType.EVENT_TIMINGS;
import static io.appium.java_client.remote.MobileCapabilityType.FULL_RESET;
import static io.appium.java_client.remote.MobileCapabilityType.NEW_COMMAND_TIMEOUT;
import static io.appium.java_client.remote.MobileCapabilityType.NO_RESET;
import static io.appium.java_client.remote.MobileCapabilityType.PLATFORM_VERSION;
import static io.appium.java_client.remote.MobileCapabilityType.UDID;
import static java.lang.String.format;
import static java.lang.System.getProperty;
import static java.util.concurrent.TimeUnit.SECONDS;
import static org.openqa.selenium.remote.CapabilityType.BROWSER_NAME;
import static org.openqa.selenium.remote.CapabilityType.PLATFORM_NAME;

import java.io.File;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.net.URL;
import java.time.Duration;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.openqa.selenium.Capabilities;
import org.openqa.selenium.NoSuchSessionException;
import org.openqa.selenium.remote.DesiredCapabilities;

import com.github.wasiqb.coteafs.appium.checker.ServerChecker;
import com.github.wasiqb.coteafs.appium.config.AppiumSetting;
import com.github.wasiqb.coteafs.appium.config.DeviceSetting;
import com.github.wasiqb.coteafs.appium.config.RecordSetting;
import com.github.wasiqb.coteafs.appium.config.android.AndroidDeviceSetting;
import com.github.wasiqb.coteafs.appium.config.enums.ApplicationType;
import com.github.wasiqb.coteafs.appium.config.enums.DeviceType;
import com.github.wasiqb.coteafs.appium.config.enums.PlatformType;
import com.github.wasiqb.coteafs.appium.config.ios.IOSDeviceSetting;
import com.github.wasiqb.coteafs.appium.error.AppiumServerStoppedError;
import com.github.wasiqb.coteafs.appium.error.DeviceAppNotFoundError;
import com.github.wasiqb.coteafs.appium.error.DeviceDesiredCapabilitiesNotSetError;
import com.github.wasiqb.coteafs.appium.error.DeviceDriverDefaultWaitError;
import com.github.wasiqb.coteafs.appium.error.DeviceDriverInitializationFailedError;
import com.github.wasiqb.coteafs.appium.error.DeviceDriverNotStartingError;
import com.github.wasiqb.coteafs.appium.error.DeviceDriverNotStoppingError;
import com.github.wasiqb.coteafs.appium.service.AppiumServer;
import com.github.wasiqb.coteafs.config.loader.ConfigLoader;
import com.google.common.reflect.TypeToken;

import io.appium.java_client.AppiumDriver;
import io.appium.java_client.MobileElement;
import io.appium.java_client.TouchAction;
import io.appium.java_client.screenrecording.BaseStartScreenRecordingOptions;
import io.appium.java_client.screenrecording.BaseStopScreenRecordingOptions;
import io.appium.java_client.screenrecording.CanRecordScreen;

/**
 * @author wasiq.bhamla
 * @param <D>
 * @param <T>
 * @since 12-Apr-2017 9:38:38 PM
 */
public abstract class Device <D extends AppiumDriver <MobileElement>, T extends TouchAction <T>> {
	private static final Logger LOG = LogManager.getLogger (Device.class);

	protected DesiredCapabilities	capabilities;
	protected D						driver;
	protected final AppiumServer	server;
	protected final DeviceSetting	setting;
	private final PlatformType		platform;

	/**
	 * @author wasiq.bhamla
	 * @param server
	 * @param name
	 * @since 13-Apr-2017 9:10:11 PM
	 */
	public Device (final AppiumServer server, final String name) {
		this.server = server;
		this.setting = ConfigLoader.settings ()
			.withKey (COTEAFS_CONFIG_KEY)
			.withDefault (COTEAFS_CONFIG_DEFAULT_FILE)
			.load (AppiumSetting.class)
			.getDevice (name);
		this.platform = this.setting.getPlatformType ();
		buildCapabilities ();
	}

	/**
	 * @author wasiq.bhamla
	 * @since Nov 17, 2017 3:34:33 PM
	 */
	public void checkServerRunning () {
		ServerChecker.checkServerRunning (this.server);
	}

	/**
	 * @author wasiqb
	 * @since Oct 28, 2018
	 * @param command
	 * @param args
	 */
	public void executeCommand (final String command, final Map <String, Object> args) {
		this.driver.executeScript (command, args);
	}

	/**
	 * @author wasiq.bhamla
	 * @since 01-May-2017 7:08:10 PM
	 * @return driver
	 */
	public D getDriver () {
		final String msg = "Getting [%s] device driver...";
		LOG.trace (format (msg, this.platform));
		return this.driver;
	}

	/**
	 * @author wasiq.bhamla
	 * @since Oct 9, 2017 4:36:02 PM
	 * @return the setting
	 */
	public DeviceSetting getSetting () {
		return this.setting;
	}

	/**
	 * @author wasiq.bhamla
	 * @since 17-Apr-2017 4:46:12 PM
	 */
	public void start () {
		startDriver ();
		setImplicitWait ();
	}

	/**
	 * @author wasiqb
	 * @since Oct 13, 2018
	 */
	public void startRecording () {
		startRecord ((CanRecordScreen) this.driver);
	}

	/**
	 * @author wasiq.bhamla
	 * @since 17-Apr-2017 4:46:02 PM
	 */
	public void stop () {
		if (this.driver != null) {
			quitApp ();
			this.driver = null;
		}
		else {
			final String message = "[%s] device driver already stopped...";
			LOG.trace (format (message, this.platform));
		}
	}

	/**
	 * @author wasiqb
	 * @since Oct 13, 2018
	 */
	public void stopRecording () {
		stopRecord ((CanRecordScreen) this.driver);
	}

	protected abstract <X extends BaseStartScreenRecordingOptions <X>> X startRecordSetting ();

	protected abstract <Y extends BaseStopScreenRecordingOptions <Y>> Y stopRecordSetting ();

	private void buildCapabilities () {
		LOG.trace ("Building Device capabilities...");
		this.capabilities = new DesiredCapabilities ();

		setCommonCapabilities ();
		setDeviceSpecificCapabilities ();

		if (this.setting.getAppType () == ApplicationType.WEB) {
			setCapability (BROWSER_NAME, this.setting.getBrowser (), this.capabilities, true);
		}
		else {
			String appPath = this.setting.getAppLocation ();
			if (appPath != null && !this.setting.isCloudApp ()) {
				String path = "%s/src/test/resources/%s";
				path = format (path, getProperty ("user.dir"), appPath);

				if (this.setting.isExternalApp ()) {
					path = appPath;
				}
				final File file = new File (path);
				if (!file.exists ()) {
					final String msg = "App not found on mentioned location [%s]...";
					LOG.error (format (msg, path));
					fail (DeviceAppNotFoundError.class, format (msg, path));
				}
				appPath = path;
			}
			if (this.setting.isCloudApp () || appPath != null) {
				setCapability (APP, appPath, this.capabilities, true);
			}
		}
		LOG.trace ("Building Device capabilities completed...");
	}

	@SuppressWarnings ("unchecked")
	private D init (final URL url, final Capabilities capability) {
		LOG.trace ("Initializing driver...");
		final TypeToken <D> token = new TypeToken <D> (getClass ()) {
			private static final long serialVersionUID = 1562415938665085306L;
		};
		final Class <D> cls = (Class <D>) token.getRawType ();
		final Class <?> [] argTypes = new Class <?> [] { URL.class, Capabilities.class };
		try {
			final Constructor <D> ctor = cls.getDeclaredConstructor (argTypes);
			return ctor.newInstance (url, capability);
		}
		catch (NoSuchMethodException | SecurityException | InstantiationException
			| IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
			fail (DeviceDriverInitializationFailedError.class,
				"Error occured while initializing device driver.", e);
		}
		return null;
	}

	private void quitApp () {
		final String message = "Closing & Quitting [%s] device driver...";
		LOG.trace (format (message, this.platform));
		try {
			this.driver.closeApp ();
			this.driver.quit ();
		}
		catch (final NoSuchSessionException e) {
			fail (AppiumServerStoppedError.class, SERVER_STOPPED, e);
		}
		catch (final Exception e) {
			fail (DeviceDriverNotStoppingError.class, "Error occured while stopping device driver.",
				e);
		}
	}

	private void setAndroidCapabilities () {
		final AndroidDeviceSetting android = this.setting.getAndroid ();
		if (android != null) {
			if (this.setting.getDeviceType () == DeviceType.SIMULATOR) {
				setCapability (AVD, android.getAvd (), this.capabilities, true);
				setCapability (AVD_READY_TIMEOUT, SECONDS.toMillis (android.getAvdReadyTimeout ()),
					this.capabilities);
				setCapability (AVD_LAUNCH_TIMEOUT,
					SECONDS.toMillis (android.getAvdLaunchTimeout ()), this.capabilities);
			}
			if (this.setting.getAppType () != ApplicationType.WEB) {
				final String packageName = android.getAppPackage ();
				final String app = this.setting.getAppLocation ();
				if (packageName == null && app == null) {
					fail (DeviceDesiredCapabilitiesNotSetError.class,
						"Either App or Package name is mandatory...");
				}
				setCapability (APP_ACTIVITY, android.getAppActivity (), this.capabilities);
				setCapability (APP_PACKAGE, android.getAppPackage (), this.capabilities);
				setCapability (APP_WAIT_ACTIVITY, android.getAppWaitActivity (), this.capabilities);
				setCapability (APP_WAIT_DURATION, android.getAppWaitTimeout (), this.capabilities);
				setCapability (APP_WAIT_PACKAGE, android.getAppWaitPackage (), this.capabilities);
				setCapability (ANDROID_INSTALL_TIMEOUT, android.getApkInstallTimeout (),
					this.capabilities);
				setCapability (AUTO_GRANT_PERMISSIONS, android.isAutoGrantPermissions (),
					this.capabilities);
			}
			else {
				setCapability (CHROMEDRIVER_EXECUTABLE, android.getChromeDriverPath (),
					this.capabilities);
			}
			setCapability (SYSTEM_PORT, android.getSystemPort (), this.capabilities);
			setCapability (ADB_PORT, android.getAdbPort (), this.capabilities);
			setCapability (DEVICE_READY_TIMEOUT, android.getDeviceReadyTimeout (),
				this.capabilities);
		}
	}

	private void setCommonCapabilities () {
		setCapability (DEVICE_NAME, this.setting.getDeviceName (), this.capabilities, true);
		setCapability (PLATFORM_NAME, this.setting.getPlatformType (), this.capabilities, true);
		setCapability (PLATFORM_VERSION, this.setting.getDeviceVersion (), this.capabilities);
		setCapability (NO_RESET, this.setting.isNoReset (), this.capabilities);
		setCapability (FULL_RESET, this.setting.isFullReset (), this.capabilities);
		setCapability (NEW_COMMAND_TIMEOUT, this.setting.getSessionTimeout (), this.capabilities);
		setCapability (CLEAR_SYSTEM_FILES, this.setting.isClearSystemFiles (), this.capabilities);
		setCapability (AUTOMATION_NAME, this.setting.getAutomationName (), this.capabilities, true);
		setCapability (UDID, this.setting.getUdid (), this.capabilities);
		setCapability (EVENT_TIMINGS, this.setting.isEventTimings (), this.capabilities);
	}

	private void setDeviceSpecificCapabilities () {
		switch (this.setting.getPlatformType ()) {
			case IOS:
				setIOSCapabilities ();
				break;
			case ANDROID:
				setAndroidCapabilities ();
				break;
			case WINDOWS:
			default:
				break;
		}
	}

	private void setImplicitWait () {
		try {
			this.driver.manage ()
				.timeouts ()
				.implicitlyWait (this.setting.getPlayback ()
					.getDefaultWait (), TimeUnit.SECONDS);
		}
		catch (final NoSuchSessionException e) {
			fail (AppiumServerStoppedError.class, SERVER_STOPPED, e);
		}
		catch (final Exception e) {
			fail (DeviceDriverDefaultWaitError.class,
				"Error occured while setting device driver default wait.", e);
		}
	}

	private void setIOSCapabilities () {
		final IOSDeviceSetting ios = this.setting.getIos ();
		if (ios != null) {
			if (this.setting.getAppType () != ApplicationType.WEB) {
				setCapability (BUNDLE_ID, ios.getBundleId (), this.capabilities, true);
			}
			setCapability (XCODE_ORG_ID, ios.getTeamId (), this.capabilities, true);
			setCapability (XCODE_SIGNING_ID, ios.getSigningId (), this.capabilities, true);
			setCapability (APP_NAME, ios.getAppName (), this.capabilities, true);
			setCapability (WDA_CONNECTION_TIMEOUT, ios.getWdaConnectionTimeout (),
				this.capabilities, true);
			setCapability ("bootstrapPath", ios.getBootstrapPath (), this.capabilities);
			setCapability ("agentPath", ios.getAgentPath (), this.capabilities);
			setCapability (UPDATE_WDA_BUNDLEID, ios.getUpdatedWdaBundleId (), this.capabilities);
			setCapability (USE_NEW_WDA, ios.isUseNewWda (), this.capabilities);
			setCapability (USE_PREBUILT_WDA, ios.isUsePrebuiltWda (), this.capabilities);
			setCapability (SHOW_XCODE_LOG, ios.isShowXcodeLog (), this.capabilities);
			setCapability (WDA_STARTUP_RETRIES, ios.getWdaStartupRetries (), this.capabilities);
			setCapability (WDA_STARTUP_RETRY_INTERVAL, ios.getWdaStartupRetryInterval (),
				this.capabilities);
			setCapability (AUTO_ACCEPT_ALERTS, ios.isAutoAcceptAlerts (), this.capabilities);
			setCapability (AUTO_DISMISS_ALERTS, ios.isAutoDismissAlerts (), this.capabilities);
			setCapability (WDA_LOCAL_PORT, ios.getWdaLocalPort (), this.capabilities);
			setCapability (LAUNCH_TIMEOUT, ios.getLaunchTimeout (), this.capabilities);
		}
	}

	private void startDriver () {
		final String msg = "Starting [%s] device driver...";
		LOG.trace (format (msg, this.platform));
		try {
			this.driver = init (this.server.getServiceUrl (), this.capabilities);
		}
		catch (final Exception e) {
			fail (DeviceDriverNotStartingError.class, "Error occured starting device driver", e);
		}
	}

	private <X extends BaseStartScreenRecordingOptions <X>> void startRecord (
		final CanRecordScreen screen) {
		final RecordSetting record = this.setting.getPlayback ()
			.getRecord ();
		if (record.isEnabled () && !this.setting.isCloudApp ()) {
			LOG.info ("Starting video recording...");
			final X option = startRecordSetting ();
			option.withTimeLimit (Duration.ofMinutes (record.getDuration ()));
			screen.startRecordingScreen (option);
		}
	}

	private void stopRecord (final CanRecordScreen screen) {
		final RecordSetting record = this.setting.getPlayback ()
			.getRecord ();
		if (record.isEnabled () && !this.setting.isCloudApp ()) {
			LOG.info ("Stopping video recording...");
			final String content = screen.stopRecordingScreen (stopRecordSetting ());
			saveRecording (content, record);
		}
	}
}